lib/tree.rb

```diff
@@ -16,40 +16,87 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n) - call stack
   def add(key, value)
-    raise NotImplementedError
+    if @root.nil?
+      @root = TreeNode.new(key, value)
+      return true
+    end
+
+    add_helper(key, value, @root)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def find(key)
-    raise NotImplementedError
+  def add_helper(key, value, current)
+    return TreeNode.new(key, value) if current.nil?
+
+    if key < current.key
+      current.left = add_helper(key, value, current.left)
+    else
+      current.right = add_helper(key, value, current.right)
+    end
+
+    return current
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def inorder
-    raise NotImplementedError
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n) - call stack
+  def find(key, current = @root)
+    return nil if current.nil?
+    return current.value if current.key == key
+
+    if key < current.key
+      return find(key, current.left)
+    else
+      return find(key, current.right)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def preorder
-    raise NotImplementedError
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def inorder(current = @root, answer = [])
+
+    if current
+      inorder(current.left, answer)
+      answer.push({key: current.key, value: current.value})
+      inorder(current.right, answer)
+    end
+
+    return answer
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def postorder
-    raise NotImplementedError
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def preorder(current = @root, answer = [])
+    if current
+      answer.push({key: current.key, value: current.value})
+      preorder(current.left, answer)
+      preorder(current.right, answer)
+    end
+
+    return answer
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def height
-    raise NotImplementedError
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def postorder(current = @root, answer = [])
+    if current
+      postorder(current.left, answer)
+      postorder(current.right, answer)
+      answer.push({key: current.key, value: current.value})
+    end
+
+    return answer
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def height(current = @root)
```

Comment:
👍 The time complexity here is right, but the space complexity is O(log n) if the tree is balanced and O(n) if it's not.